near/src/runtime.rs

```diff
@@ -854,66 +853,42 @@ impl RuntimeAdapter for NightshadeRuntime {
-            cursor.write_u32::<LittleEndian>(value.len() as u32)?;
-            cursor.write_all(value.as_ref())?;
-        }
-        // TODO(1048): Save on disk an snapshot, split into chunks and compressed. Send chunks instead of single blob.
-        debug!(target: "runtime", "Read state part #{} for shard #{} @ {}, size = {}", part_id, shard_id, state_root.hash, result.len());
-        // TODO add proof in Nightshade Runtime
-        Ok((StatePart { shard_id, part_id, data: result }, vec![]))
+        part_id: u64,
+        num_parts: u64,
+    ) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
+        let data = self
+            .trie
+            .get_trie_nodes_for_part(part_id, num_parts, state_root)?
+            .try_to_vec()
+            .expect("Failed to serialize state part");
+        Ok(data)
     }
 
-    fn accept_state_part(
+    fn validate_state_part(
         &self,
         state_root: &StateRoot,
-        part: &StatePart,
-        _proof: &Vec<u8>,
+        part_id: u64,
+        num_parts: u64,
+        data: &Vec<u8>,
     ) -> Result<(), Box<dyn std::error::Error>> {
-        if part.part_id > 0 {
-            /* TODO MOO */
-            return Ok(());
-        }
-        debug!(target: "runtime", "Writing state part #{} for shard #{} @ {}, size = {}", part.part_id, part.shard_id, state_root.hash, part.data.len());
-        // TODO prove that the part is valid
-        let mut state_update = TrieUpdate::new(self.trie.clone(), CryptoHash::default());
-        let state_part_len = part.data.len();
-        let mut cursor = Cursor::new(part.data.clone());
-        while cursor.position() < state_part_len as u64 {
-            let key_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut key = vec![0; key_len];
-            cursor.read_exact(&mut key)?;
-            let value_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut value = vec![0; value_len];
-            cursor.read_exact(&mut value)?;
-            state_update.set(key, DBValue::from_slice(&value));
-        }
-        let (store_update, root) = state_update.finalize()?.into(self.trie.clone())?;
-        if root != state_root.hash {
-            return Err("Invalid state root".into());
-        }
-        store_update.commit()?;
-        Ok(())
+        let trie_nodes: PartialStateStruct =
+            BorshDeserialize::try_from_slice(data).expect("Failed to deserialize state part");
+        Ok(Trie::validate_trie_nodes_for_part(state_root, part_id, num_parts, &trie_nodes)?)
     }
 
-    fn confirm_state(&self, _state_root: &StateRoot) -> Result<bool, Error> {
-        // TODO(1457): approve that all parts are here
-        Ok(true)
+    fn confirm_state(&self, state_root: &StateRoot, data: &Vec<Vec<u8>>) -> Result<(), Error> {
+        let mut parts = vec![];
+        for part in data {
+            parts.push(
+                BorshDeserialize::try_from_slice(part).expect("Failed to deserialize state part"),
```

Comment:
should be `.expect("Part was already validated earlier, so could never fail here"),`
For errors below as well, every expect/unwrap should have an explanation when it can fail and why it's ok to panic

Reply:
++